### ISSUE TITLE
Compilation removes needed IF statements

### ISSUE THREAD

**Issue by amwmedia** [rollup/rollup#2326]
This was a weird one to track down, but [here's the REPL link](https://rollupjs.org/repl?version=0.62.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiolMjBUUkVFLVNIQUtJTkclMjAqJTJGJTVDbmltcG9ydCUyMCcuJTJGbWF0aHMuanMnJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1hdGhzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMndpbmRvdy50ZXN0JTIwJTNEJTIwKGZ1bmN0aW9uJTIwKCklMjAlN0IlNUNuJTVDdHZhciUyMGNvbmZpZyUyQyUyMHByb3BzJTIwJTNEJTIwJTdCJTdEJTNCJTVDbiU1Q24lNUN0ZnVuY3Rpb24lMjBkb1VwZGF0ZSUyMChmb3JjZSklMjAlN0IlNUNuJTVDdCU1Q3R2YXIlMjBjbHMlM0IlNUNuJTVDbiU1Q3QlNUN0Zm9yJTIwKGNscyUyMGluJTIwY29uZmlnKSUyMCU3QiU1Q24lNUN0JTVDdCU1Q3RpZiUyMChwcm9wcyU1QmNscyU1RCUyMCUzRCUzRCUyMG51bGwpJTIwJTdCJTIwY29udGludWUlM0IlMjAlN0QlNUNuJTVDdCU1Q3QlNUN0JTJGJTJGJTVFJTVFJTVFJTVFJTVFJTVFJTVFJTVFJTVFJTIwV0hBVCUyMElOJTIwVEhFJTIwV09STEQhJTNGJTVDbiU1Q3QlNUN0JTdEJTVDbiU1Q3QlN0QlNUNuJTVDbiU1Q3RyZXR1cm4lMjAlN0IlMjBkb1VwZGF0ZSUyMCU3RCUzQiU1Q24lN0QoKSklM0IlNUNuJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmlpZmUlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==).

As you can see, the compiler is removing the if statement, but leaving in the curly braces. Don't pick on the example too much, I reduced it as much as I could to get it down to something that demonstrates the core issue.

**Comment by lukastaegert** [comment]
Sorry for the issues and thanks for tracking it down. We are aware that there are some issues with the tree-shaking in the latest version for which a fix is already in preparation. Could you please verify if installing rollup via `npm i rollup/rollup#tree-shaking-fixes` solves the issue for you?

**Comment by lukastaegert** [comment]
It should be noted though that in your example, there is no mechanism for either `props` or `config` to be anything but an empty object.

**Comment by lukastaegert** [comment]
The curly braces are left behind because rollup does not yet have a mechanism to merge nested scopes i.e. the if-statement is (wrongly) simplified but as rollup does not detect yet if there are nested declarations, the curly braces are kept to avoid name conflicts.

**Comment by amwmedia** [comment]
when I install via that tag I get a compilation error `Cannot find module 'rollup'`. Not sure why.

I agree that the example is not perfect. the actual code DOES have another exposed method that fills the objects that start out empty, however I haven't been able to figure out a simplified version of that to demonstrate the issue. I'll post here if I do.

**Comment by lukastaegert** [comment]
You are probably using an old version of npm if the direct install does not work. You need npm 5+ for the install script to work on direct Github installation.

**Comment by amwmedia** [comment]
I'm on npm 5.6.0. Also tried with yarn 1.7.0.

**Comment by lukastaegert** [comment]
Interesting. Are you in the same directory as your node_modules folder when running the install command? I just checked again locally and it works for me both for the CLI and the node API but your setup may be more complicated than mine.

**Comment by lukastaegert** [comment]
Anyway I manually checked your REPL link and it seems to correctly keep the if-statement so I will mark your issue as being resolved by #2315.

**Comment by EricSch** [comment]
@lukastaegert I had a similar issue, but couldn't reproduce it with a REPL. The code looks like that (I shortend the functions bodys):
```
var animationTypes = {};
/* tslint:disable:object-literal-sort-keys */
animationTypes.expand = function (direction) {
    var scale = direction === UP || direction === DOWN ? 'scaleY' : 'scaleX';
    return {
        start: { transform: scale + "(" + startScale + ")", transformOrigin: origin },
        end: { transform: scale + "(" + endScale + ")" }
    };
};
animationTypes.slide = function (direction) {
    var translate = direction === LEFT || direction === RIGHT ? 'translateX' : 'translateY';
    var start = direction === RIGHT || direction === DOWN ? -100 : 100;
    var end = 0;
    return {
        start: { transform: translate + "(" + start + "%)" },
        end: { transform: translate + "(" + end + "%)" }
    };
};
/**
 * @hidden
 */
var AnimationService = /** @class */ (function () {
    function AnimationService(animationBuilder) {
        this.animationBuilder = animationBuilder;
        this.start = new EventEmitter();
        this.end = new EventEmitter();
    }
    AnimationService.prototype.play = function (element, options, flip) {
        if (!this.flip || this.flip.horizontal !== flip.horizontal ||
            this.flip.vertical !== flip.vertical) {
            this.flip = flip;
            var type = options.type || DEFAULT_TYPE;
            var statesFn = animationTypes[type];
            if (statesFn) {
                var direction = this.getDirection(flip, options);
                var states = statesFn(direction);
                this.playStates(element, states, options);
            }
            else if (isDevMode()) {
                throw new Error("Unsupported animation type: \"" + type + "\". The supported types are slide, expand, fade and zoom.");
            }
        }
    };
    AnimationService.prototype.ngOnDestroy = function () {
        this.stopPlayer();
    };
```


The output in the bundle was:
```
var DEFAULT_TYPE = 'slide';
var DEFAULT_DURATION = 100;
/**
 * @hidden
 */
var AnimationService$1 = /** @class */ (function () {
    function AnimationService(animationBuilder) {
        this.animationBuilder = animationBuilder;
        this.start = new EventEmitter();
        this.end = new EventEmitter();
    }
    AnimationService.prototype.play = function (element, options, flip) {
        if (!this.flip || this.flip.horizontal !== flip.horizontal ||
            this.flip.vertical !== flip.vertical) {
            this.flip = flip;
            var type = options.type || DEFAULT_TYPE;
            if (isDevMode()) {
                throw new Error("Unsupported animation type: \"" + type + "\". The supported types are slide, expand, fade and zoom.");
            }
        }
    };
    AnimationService.prototype.ngOnDestroy = function () {
        this.stopPlayer();
    };

```
The whole `if` and the var `animationTypes` was gone. After using your `npm i rollup/rollup#tree-shaking-fixes` the bundles seems correct again.

Do you have any date for this release? Or which version I could use to avoid that (using .62)?
Thanks

**Comment by lukastaegert** [comment]
I hope we can release soon but I cannot give you a definite date as significant parts of our very small core team are on vacation at the moment and the (unfortunately somewhat extensive) fix has not been reviewed yet.

Until then, I would recommend to try 0.60.7 first (which does not yet have the "missing property===undefined" logic) and if that does not work, try 0.58.2 which should definitely work as it does not yet contain any of the new conditional evaluation logic.

If it works for you, you could also directly install from the fix branch via `npm i rollup/rollup#tree-shaking-fixes`. Not sure if/how this works with yarn, though.

**Comment by lukastaegert** [comment]
Fix released via Rollup v0.63.0, thanks for being patient!
